File: 도전 문제/10주차_분할 정복/1074_Z/skyqnaqna.swift
/*
 백준 1074 Z
 21.08.23
 https://github.com/skyqnaqna/algorithm_study
 */

import Foundation

let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = input[0], r = input[1], c = input[2]

func recursion(_ n: Int, _ r: Int, _ c: Int) -> Int {
  if n == 0 { return 0 } // n이 0이면 크기가 1인 사각형이므로 0번째임

  let half = 1 << (n - 1) // 사각형 한변의 길이
  /*
   [1] = half x half
        |
    [1] | [2]
  ------+------
    [3] | [4]
        |
   */
  // [1]에 위치하면 2^(n-1) * 2^(n-1) 크기인 사각형 안에서 다시 탐색
  if r < half && c < half { return recursion(n - 1, r, c) }
  // [2]에 위치하면 [1]의 크기인 half * half 만큼 더하고 [2]에서 다시 탐색
  else if r < half && c >= half { return half * half + recursion(n - 1, r, c - half) }
  // [3]에 위치하면 half * half 크기가 이미 두번 지나왔으므로 그만큼 더해주고 탐색
  else if r >= half && c < half { return half * half * 2 + recursion(n - 1, r - half, c) }
  // [4]에 위치하면 마찬가지로 누적된 개수 더해주고 탐색
  else { return half * half * 3 + recursion(n - 1, r - half, c - half) }
}

print(recursion(n, r, c))

File: 기본 문제/11주차_트리/5639_이진 검색 트리/skyqnaqna.cpp
/*
 백준 5639 이진 검색 트리
 21.08.25
 https://github.com/skyqnaqna/algorithm_study
 */

#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using dd = double;
using pii = pair<int, int>;
using pll = pair<ll, ll>;

#define INF (int)2e9
#define endl "\n"

vector <int> tree;

void pre_to_post(int s, int e) {
    if (s > e) return; // left or right 없는 경우
    if (s == e) { // 리프 노드인 경우
        cout << tree[s] << endl;
        return;
    }

    int root = tree[s];
    int left = s + 1, right = e + 1;

    for (int i = s + 1; i <= e; ++i) {
        if (root < tree[i]) {
            right = i;
            break;
        }
    }

    // root보다 작은 요소들은 left 서브 트리, 큰 요소들은 right 서브 트리로 나눈다
    pre_to_post(left, right - 1);
    pre_to_post(right, e);
    cout << root << endl; // root를 마지막에 출력하여 후위 순회 탐색
}

int main() {
    ios::sync_with_stdio(0); cin.tie(0);
    freopen("input.txt", "r", stdin);

    int n;
    while (cin >> n) {
        tree.push_back(n);
    }

    pre_to_post(0, tree.size() - 1);

    return 0;
}